### DISCUSSION TITLE
Creating view failing with "Expression does not eval to a function. " error

### DISCUSSION THREAD

**Discussion by oraclevet** [apache/couchdb#2933] [Q&A] [answered]
Creating a view(map-reduce) giving  "Expression does not eval to a function. " error.

## Description

Creating a view.

## Steps to Reproduce

code to create the view is as below,

function (doc) {
  // first extract what we want to use
  const vendors = doc.cve &&
    doc.cve.affects &&
    doc.cve.affects.vendor &&
    doc.cve.affects.vendor.vendor_data;
  // since we can't be sure vendor_data does exist
  // let's check if it's really an array
  if (Array.isArray(vendors) === false)
    return; // if not just return and proceed with the next document
  // iterate through every "vendor" in that array
  vendors.forEach(vendor => {
    // let's remember the vendor name
    const vendorName = vendor.vendor_name;
    // let us also unwrap the product_data
    const products = vendor.product &&
      vendor.product.product_data;
    // again, if products is no array
    if (Array.isArray(products) === false)
      return; // return and proceed with the next vendor
    products.forEach(product => {
        // now emit one vendor/product-pair
        emit(vendorName, product.product_name);
    })
  })
}

## Expected Behaviour

view created.

## Your Environment

{"couchdb":"Welcome","version":"3.1.0","git_sha":"ff0feea20","uuid":"b401bbbf089d0d3452eaad6afc60a8f5","features":["access-ready","partitioned","pluggable-storage-engines","reshard","scheduler"],"vendor":{"name":"The Apache Software Foundation"}}

* CouchDB version used: 3.1.0
* Browser name and version: Chrome/Firefox
* Operating system and version: Ubuntu 18.04 running under WSL.

## Additional Context

[TIP]:  # ( Add any other context about the problem here. )

**Comment by wohali** [comment, marked as answer]
Ubuntu 18.04 is built against SpiderMonkey 1.8.5, meaning you cannot use the arrow anonymous function declarations that you're using. We are unable to supply SpiderMonkey 60 or 68 for Ubuntu 18.04 at this time.

To keep your arrow-function anonymous syntax, you would need to move to Debian Buster, Ubuntu 20.04, or CentOS 8, or use our Docker image, which is based on Debian Buster.

You can also just rewrite your function using a `let myfn = function() { ... };` syntax, then call that using `myfn()` in your `foreach` statements.

Be sure to mark this answer as accepted if it resolves your issue!

**Comment by florianb** [comment]
Thank you @wohali - i will guide @oraclevet through this.

We had a hard night on this.. :smile:
